odoo/addons/base/models/ir_attachment.py

```diff
@@ -475,6 +463,23 @@ def check(self, mode, values=None):
             records.check_access_rights(access_mode)
             records.check_access_rule(access_mode)
 
+    def _collect_records_to_check(self, mode):
```

Comment:
Hello @schoffeniels 

Had a look at the new implementation, it looks better!

Had a few comments to try to improve / polish.

Ideally we want to avoid copy/pasting a large portion of the code when re-defining this in the knowledge module.

I was thinking more something like the following.

In the check method:

```py
# ...
records_to_check = tuple()
if self:
    # DLE P173: `test_01_portal_attachment`
    self.env['ir.attachment'].flush_model(['res_model', 'res_id', 'create_uid', 'public', 'res_field']) 
    records_to_check = self._cr.execute('SELECT res_model, res_id, create_uid, public, res_field FROM ir_attachment WHERE id IN %s', [tuple(self.ids)])

model_ids = self._collect_records_to_check(records_to_check, mode)
# ...
```

Then the method itself:

```py
def _collect_records_to_check(self, records_to_check, mode):
    model_ids = defaultdict(set)            # {model_name: set(ids)}
    for res_model, res_id, create_uid, public, res_field in self._cr.fetchall():
        if public and mode == 'read':
            continue
        if not self.env.is_system() and (res_field or (not res_id and create_uid != self.env.uid)):
            raise AccessError(_("Sorry, you are not allowed to access this document."))
        if not (res_model and res_id):
            continue
        model_ids[res_model].add(res_id)
    return model_ids
```

Then in the knowledge module:

```py
def _collect_records_to_check(self, records_to_check, mode):
    """ Custom behavior for the Knowledge module, we assume that read access to the model is sufficient to
    be able to read the articles covers.
    This is necessary as those are stored as an attachment with res_id 0, and the base implementation of this method
    considers that you have to be the creator of such an attachment to be able to access it.
    This override allows implementing a 'bank of images' for Knowledge, using the same attachment for multiple articles. """

    if mode == 'read' and any([
        res_model == 'knowledge.article' and res_field == 'cover_image_id'
        for res_model, res_id, create_uid, public, res_field in records_to_check]):
        if not self.env['knowledge.article'].check_access_rights('read'):
           raise AccessError(_("Sorry, you are not allowed to access this document."))

        records_to_check = filter(records_to_check, lambda res_model, res_id, create_uid, public, res_field:
            res_model != 'knowledge.article' or res_field != 'cover_image_id')
    
    return super()._collect_records_to_check(records_to_check, mode)
```

I just typed this into github so there might be typos / errors, but it should give you a general idea of the direction I had in mind.

Your current implementation is different as it checks if the attachment is used for a cover, and if the current user has access to it.
This is "better" as it would hide covers used for private articles for example.
However, your current implementation should be improved to accumulate records to check and then run:
```
records.check_access_rights(access_mode)
records.check_access_rule(access_mode)
```

At the end, on all records at once, similarly to the base "check" implementation.
Indeed, running a search_count request inside a loop is what we call a "red flag" in terms of performances.

However, we might want to discuss if checking read access is enough, as I doubt people would store anything sensitive in an article cover :thinking: 
That way it's simpler and more performant.

Thank you for all your time and investigations on this one!
I feel like we are getting very close to a robust solution.

Cheers!